src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,828 @@
+  // The last guess is to just say "c++"
+  return std::string("c++");
+}
+
+std::string get_cxx_std(const std::string& cxx) {
+  const char* standards[] = { "-std=c++17", "-std=c++14" };
+  int fd = ::open("/tmp/.hail_empty.cc", O_WRONLY|O_CREAT|O_TRUNC, 0666);
+  if (fd >= 0) ::close(fd);
+  for (int j = 0; j < 2; ++j) {
+    std::stringstream ss;
+    ss << cxx << " " << standards[j] << " /tmp/.hail_empty.cc 2>1";
+    auto s = run_shell_get_first_line(ss.str().c_str());
+    if (!strstr(s.c_str(), "unrecognized") && !strstr(s.c_str(), "invalid")) {
+      return standards[j];
+    }
+  }
+  // Default to pass "-std=c++11", a minimum requirement
+  return "-std=c++11";
+}
+
+std::string get_module_dir() {
+  // This gives us a distinct temp directory for each process, so that
+  // we can manage synchronization between different of the files in the
+  // temp directory using only in-memory std::mutex'es, rather than
+  // dealing with complexities of file locking.
+  char buf[512];
+  strcpy(buf, "/tmp/hail_XXXXXX");
+  return ::mkdtemp(buf);
+}
+
+std::string get_perl_name() {
+  std::string name("/usr/bin/perl");
+  if (file_exists(name)) return name;
+  name = run_shell_get_first_line("which perl 2>/dev/null");
+  if (strstr(name.c_str(), "perl")) return name;
+  // The last guess is to just say "perl"
+  return std::string("perl");
+}
+
+class ModuleConfig {
+ public:
+  bool is_darwin_;
+  std::string ext_cpp_;
+  std::string ext_lib_;
+  std::string ext_mak_;
+  std::string ext_new_;
+  std::string module_dir_;
+  std::string cxx_name_;
+  std::string cxx_std_;
+  std::string java_home_;
+  std::string java_md_;
+  std::string perl_name_;
+
+ public:
+  ModuleConfig() :
+#if defined(__APPLE__) && defined(__MACH__)
+    is_darwin_(true),
+#else
+    is_darwin_(false),
+#endif
+    ext_cpp_(".cpp"),
+    ext_lib_(is_darwin_ ? ".dylib" : ".so"),
+    ext_mak_(".mak"),
+    ext_new_(".new"),
+    module_dir_(get_module_dir()),
+    cxx_name_(get_cxx_name()),
+    cxx_std_(get_cxx_std(cxx_name_)),
+    java_home_(get_java_home()),
+    java_md_(is_darwin_ ? "darwin" : "linux"),
+    perl_name_(get_perl_name()) {
+  }
+  
+  std::string get_lock_name(const std::string& key) {
+    std::stringstream ss;
+    ss << module_dir_ << "/hm_" << key << ".lock";
+    return ss.str();
+  }
+  
+  std::string get_lib_name(const std::string& key) {
+    std:: stringstream ss;
+    ss << module_dir_ << "/hm_" << key << ext_lib_;
+    return ss.str();
+  }
+  
+  std::string get_new_name(const std::string& key) {
+    std:: stringstream ss;
+    ss << module_dir_ << "/hm_" << key << ext_new_;
+    return ss.str();
+  }
+  
+  void ensure_module_dir_exists() {
+    int rc = ::access(module_dir_.c_str(), R_OK);
+    if (rc < 0) { // create it
+      rc = ::mkdir(module_dir_.c_str(), 0666);
+      if (rc < 0) perror(module_dir_.c_str());
+      rc = ::chmod(module_dir_.c_str(), 0755);
+    }
+  }
+};
+
+ModuleConfig config;
+
+class ModuleCache {
```

Comment:
Please remove the cache.

Reply:
Sorry, that was over-zealous, leave it.  Obviously, we don't want the build directory to grow without bound (although I think it'll be a while before that is a danger).

Reply:
The motivation for the ModuleCache is based on observing the behavior during
shuffle's, where it currently generates a new PackDecoder for each item received,
which - prior- to adding the ModuleCache - used to be constructing a new NativeModule.
So we can get much large numbers of NativeModule constructor/destructor calls than
I had originally expected.

There may be higher-level ways to address that, but the ModuleCache seemed a simple
way to minimize the churn of NativeModule construction/destruction in this case.

Reply:
> where it currently generates a new PackDecoder for each item received

This might be a sign something is wrong.  The current code is designed to create all these things once per partition, amortizing the overhead over rows within the partition.  I gave a quick scan of the code and the current (non-C++) code path shouldn't be doing that (although if it is, there is a bug and it should get fixed).

Hmm.  Caching across partitions should happen for free because in general multiple partitions will be running in an executor simultaneously, each with a shared_ptr to the module.  So I don't think there gives any benefit if we fix the above problem.

Anyway, leave it in for now and but let's make sure to go back and investigate/squash any per -item activity of this magnitude.

Reply:
Yes, and I think that may not be the only thing wrong in shuffle.  While I was running it with
printf's, I saw what I thought looked like a succession of RegionValue's getting allocated in the same Region without any reset/clear.  I haven't pursued that yet because I was focused on debugging the RowStore, but it's on my list to go back and investigate/confirm/refute.

The multiple PackDecoders was specifically happening in shuffle, where IIRC it gets each
item as an Array[Byte], and then builds some kinds of InputStream to access the Array[Byte],
and then builds the PackDecoder on top of the stream.  The C++ decoder isn't taking any different path at that level, so I believe it's inherent to the current design, where the 
MatrixTable format has a sequence of byte=01, RV, byte=01, RV, ..., byte=00 - but the
shuffle data (I think) is a sequence of Array[Byte] managed by Spark.  Without an interface to point an existing PackDecoder at a new stream, or to push more data onto the stream for an existing PackDecoder, it sdseems it has to create a new one each time.

Reply:
"Caching across partitions should happen for free because in general multiple partitions will be running in an executor simultaneously, each with a shared_ptr"

Yes, that is one benefit of the ModuleCache - but before the ModuleCache was put it in,
each worker would get its own separate NativeModule rather than having any sharing at all.  Which also worked ok, but sharing seemed both more natural and more efficient, even if it isn't exactly "for free".

Reply:
You always get complete sharing of the DLL file, but without the cache you never get any
sharing of the C++ NativeModule objects.  Each executor
gets the (moduleKey, moduleBinary) pair, but then has to use those to get hold of a
NativeModule.  With the cache, they all get pointers to the same NativeModule object; without the cache, each points to their own NativeModule object.

All those NativeModule's are referring to the same DLL file.  You *must* have only one DLL file (to avoid redefining symbols).  And regardless of the number of NativeModule objects,
the DLL file stays around for re-use and you will only build it precisely once in a session.

I think you are confused about the relationship between worker NativeModule's, and the DLL files.  A worker NativeModule (on the C++ side) represents a user of the DLL (or multiple Scala-side users, in the cached/shared case); there can be multiple users of the DLL file, or no users of the DLL file (as naturally occurs between the master building it, and the first worker trying to use it), without affecting the existence of the DLL files. Like open file descriptors referring to a file.

Reply:
Ah, yes, I didn't realize the cache was also doing the interning.  Clearly we want that.  How about a `std::unordered_map<std::string, std::weak_ptr<NativeModule>>`?  Dead simple.

Reply:
OK. Would churn in the case of single partition and new-decoder-per-item (and when churning, I haven't measured the performance impact of doing dlopen/dlclose many times) but if I understand correctly your intent is to make high-level changes to avoid that.

I'll proceed with the weak_ptr approach on that assumption.